README.md
# An iOS Wanikani Leech Widget implemented using Scriptable
![IMG_9363|690x467](https://res.cloudinary.com/mca62511/image/upload/v1632987632/IMG_9565_1_ie4my7.png)

If you already know what leeches are, and you already know what Scriptable is, then you can go ahead and grab the script from [here](https://github.com/mcaubrey/ios-wanikani-leech-widget-scriptable/blob/main/script.js). I think it should be fairly self explanatory to anyone whose added an iOS widget using Scriptable before. 

# Latest Update: September 30, 2021

Changed to use the new Shin Wanikani Leech Trainer API, and switched from scraping to using the official API. This should make it more reliable overall, as well as potentially solve the problem that some people were having where the shown item did not match the shown meaning. 

# What Are Leeches? 

The term "leech" comes from Anki. Or at least I think it does. Either way, the [Anki manual](https://docs.ankiweb.net/leeches.html) has a [good explanation](https://docs.ankiweb.net/leeches.html) of what they are. In short, leeches are radicals, kanji and vocabulary that for whatever reason you just can't seem to remember. They're the ones that you fail over and over again.

In Anki, there are various ways of dealing with this. For example, you can opt to suspend leeches automatically so that they don't get in the way of your progress.

In Wanikani, we have no choice but to squash (learn) those leeches.

Towards that effort (and also because I wanted a Wanikani related widget on my home screen) I threw together a widget that displays leeches so that you can be reminded of them every time you look at your phone.

# What Is Scriptable?

[Scriptable](https://apps.apple.com/us/app/scriptable/id1405459188) is an automation app for iOS. It lets you write JavaScript scripts that interact with native features of iOS. You can also use it to create home screen widgets using JavaScript. 

# Getting Started

There are two ways to you can go about installing the script: Downloading it, or copy/pasting it. Choose the way that is easier for you.

## Install By Downloading To iCloud

First install the [Scriptable](https://apps.apple.com/us/app/scriptable/id1405459188) app. Open it so that it will create a folder within your iCloud folder.

From there, download [this JavaScript file](https://raw.githubusercontent.com/mcaubrey/ios-wanikani-leech-widget-scriptable/main/Wanikani%20Leeches.js) and place it inside `iCloud Drive > Scriptable`. 

After it is downloaded, you should be able to see it from within the Scriptable app on the Scripts tab.

## Install By Copy/Pasting Into The App

If the download method failed for some reason, you can copy and paste the script directly into the app. Just open [this JavaScript file](https://raw.githubusercontent.com/mcaubrey/ios-wanikani-leech-widget-scriptable/main/Wanikani%20Leeches.js) and copy the contents. Open the Scriptable app, click on the icon on the top right corner, and then paste the contents of the script.

After it is copied,  you should be able to see it from within the Scriptable app on the Scripts tab. You will need to manually change the name of the script (by default it will be "Untitled Script").

## Adding The Widget

Add the Scriptable widget just like you would any other widget. When the widget is initially added to the home screen it will say "Select script in widget configuration." While the home screen is still in edit mode, click on the widget again to bring up the widget configuration dialog.

You'll see three options that you can set: "Script," "When Interacting," and "Parameter." Click on "Script" and select the name of the Wanikani Leech widget script. If you followed the download method above, it will probably be called "Wanikani Leeches."

In the "Parameter" field, add your Wanikani API token.

And after that you should be set!

Here is a video example in case that was hard to follow.

![ezgif.com-gif-maker (1)|231x500](https://res.cloudinary.com/mca62511/image/upload/v1632580650/ezgif.com-gif-maker_1_ly08k5.gif)


# How It Works

The leeches are being pulled from the API used by the [Shin Wanikani Leech Trainer](https://community.wanikani.com/t/userscript-leech-training/36978) Userscript by Ross Hendry. From there I hit the official Wanikani API to grab details about the items. 

# Known Issues

 * In situations of low internet connectivity, it will often display as "Oops, something went wrong."

CHANGELOG.md
## [0.0.4] - January 30, 2022

* Update URL used to access the Shin Wanikani Leech Trainer API

## [0.0.3] - September 30, 2021

* Switch to the new Shin Wanikani Leech Trainer API
* Switch to using the official API instead of scraping. 
* Fixed a problem where due to scraping issues, sometimes the shown item and the meaning did not match. 

## [0.0.2] - September 25, 2021

* Add a link to the Wanikani forums in the comments.

## [0.0.1] - September 25, 2021

* Initial release.


Wanikani Leeches.js
// Variables used by Scriptable.
// These must be at the very top of the file. Do not edit.
// icon-color: pink; icon-glyph: torii-gate;

/* 
  Leech Widget version 0.0.4
  Join the conversation on the Wanikani community forums!
  https://community.wanikani.com/t/scriptable-ios-wanikani-leeches-widget/53682
*/

/*
  You can hard code your API Token here if you'd like, though
  it would probably be better to add your token as the "Parameter"
  in the widget settings.
*/
const apiKey = args.widgetParameter || '';

// Change default font sizes!
const bigSize = 42;
const smallSize = 16;
const spacerSize = 5;

// Change default colors!
const vocabColor = "#aa00fe";
const vocabShadowColor = "#9300dd";
const vocabTextColor = "#ffffff";
const kanjiColor = "#ff01aa";
const kanjiShadowColor = "#dd0093";
const kanjiTextColor = "#ffffff";

/*
  +-------------------------------------------+
  |         ONLY EDIT THE CODE BELOW          |
  |      IF YOU KNOW WHAT YOU'RE DOING!       |
  +-------------------------------------------+
*/

let widget = await createWidget();

if (config.runsInWidget) {
  Script.setWidget(widget);
} else {
  widget.presentSmall();
}
Script.complete();

async function createWidget() {
  if (apiKey === '') {
    return makeApiTokenNotice();
  }

  let r = new Request(`https://wk-leeches.herokuapp.com/leeches`);
  r.headers = { "Authorization": "Bearer " + apiKey };
  const json = await r.loadJSON()

  let listwidget = false
  let iterations = 0;
  do {
    if (iterations >= 1) {
      return makeErrorNotice();
    }

    listwidget = await attemptWidgetCreation(json)
    iterations++
  } while (listwidget === false)

  return listwidget;
}

async function attemptWidgetCreation(json) {
  try {
    let listwidget = new ListWidget();
    listwidget.url = "https://wanikani.com"

    if (json.leeches.length === 0) {
      return makeNoLeechNotice();
    }

    const index = Math.floor(Math.random() * (json.leeches.length - 1));
    const type = json.leeches[index].type

    switch (type) {
      case "kanji":
        listwidget.backgroundColor = new Color(kanjiColor);
        await buildLeechWidget(listwidget, type, json.leeches[index]);
        break;
      case "vocabulary":
        listwidget.backgroundColor = new Color(vocabColor);
        await buildLeechWidget(listwidget, type, json.leeches[index]);
        break;
      default:
    }

    return listwidget
  } catch (e) {
    console.log(e);
    throw e;
    return false;
  }
}

function makeApiTokenNotice() {
  let listwidget = new ListWidget();
  listwidget.backgroundColor = new Color("#ffaa01");

  let notice = listwidget.addText("API Token is not set.");
  notice.centerAlignText();
  notice.font = Font.blackRoundedSystemFont(smallSize)
  notice.textColor = new Color("#ffffff");
  shadow(notice, smallSize, "#dd9200");

  return listwidget;
}

function makeNoLeechNotice() {
  const color = "#008343";
  const shadowColor = "#006333";
  let listwidget = new ListWidget();
  listwidget.backgroundColor = new Color(color);

  let notice = listwidget.addText("Good job genius! ❤︎ You have no leeches!");
  notice.centerAlignText();
  notice.font = Font.blackRoundedSystemFont(smallSize)
  notice.textColor = new Color("#ffffff");
  shadow(notice, smallSize, shadowColor);

  return listwidget;
}

function makeErrorNotice() {
  let listwidget = new ListWidget();
  listwidget.backgroundColor = new Color("#ee2201");

  let notice = listwidget.addText("Oops! Something went wrong...");
  notice.centerAlignText();
  notice.font = Font.blackRoundedSystemFont(smallSize)
  notice.textColor = new Color("#ffffff");
  shadow(notice, smallSize, "#922201");

  return listwidget;
}

async function buildLeechWidget(listwidget, type, json) {
  const subject_id = json.subject_id
  const { reading, meaning } = await getSubjectInfo(type, subject_id);
  let textColor = vocabTextColor;
  let shadowColor = vocabShadowColor;
  if (type === "kanji") {
    textColor = kanjiTextColor;
    shadowColor = kanjiShadowColor;
  }

  let fSize = bigSize;
  if (json.name.length !== 1) fSize = bigSize - ((json.name.length) * 4);
  if (fSize < smallSize) fSize = smallSize;
  let target = listwidget.addText(json.name);
  target.centerAlignText();
  target.font = Font.blackRoundedSystemFont(fSize)
  target.textColor = new Color(textColor);
  shadow(target, fSize, shadowColor);

  listwidget.addSpacer(spacerSize);

  let rt = listwidget.addText(reading);
  rt.centerAlignText();
  rt.font = Font.lightSystemFont(smallSize);
  rt.textColor = new Color(textColor);
  let mt = listwidget.addText(meaning);
  mt.centerAlignText();
  mt.font = Font.lightSystemFont(smallSize);
  mt.textColor = new Color(textColor);
  shadow(mt, smallSize, shadowColor);
  shadow(rt, smallSize, shadowColor);
}

async function getSubjectInfo(type, subject_id) {
  const url = 'https://api.wanikani.com/v2/subjects/' + subject_id;

  let r = new Request(url);
  r.headers = { "Authorization": "Bearer " + apiKey };
  const result = await r.loadJSON()

  const meaning = result.data.meanings.filter(a => a.primary)[0].meaning;
  const reading = result.data.readings.filter(a => a.primary)[0].reading;

  return {
    meaning,
    reading
  }
}

function shadow(target, size, color) {
  p = size * (3 / 32);

  target.shadowColor = new Color(color);
  target.shadowOffset = new Point(p, p);
  target.shadowRadius = 0.1;
}

script.js
// Variables used by Scriptable.
// These must be at the very top of the file. Do not edit.
// icon-color: pink; icon-glyph: torii-gate;

/* 
  Leech Widget version 0.0.1
  Join the conversation on the Wanikani community forums...
  ...theoretically. I plan on sharing this on the Wanikani
  community forums, and when I do I'll put a link here.
*/

/*
  You can hard code your API Token here if you'd like, though
  it would probably be better to add your token as the "Parameter"
  in the widget settings.
*/
const apiKey = args.widgetParameter || ''

// Change default font sizes!
const bigSize = 42;
const smallSize = 16;
const spacerSize = 5;

// Change default colors!
const vocabColor = "#aa00fe";
const vocabShadowColor = "#9300dd";
const vocabTextColor = "#ffffff";
const kanjiColor = "#ff01aa";
const kanjiShadowColor = "#dd0093";
const kanjiTextColor = "#ffffff";
const radicalColor = "#01aaff";
const radicalShadowColor = "#0093dd";
const radicalTextColor = "#ffffff"

/*
  +-------------------------------------------+
  |         ONLY EDIT THE CODE BELOW          |
  |      IF YOU KNOW WHAT YOU'RE DOING!       |
  +-------------------------------------------+
*/

let widget = await createWidget();

if (config.runsInWidget) {
  Script.setWidget(widget);
} else {
  widget.presentSmall();
}
Script.complete();

async function createWidget() {
  if (apiKey === '') {
    return makeApiTokenNotice();
  }

  let r = new Request(`https://wk-stats.herokuapp.com/leeches/lesson?api_key=${apiKey}`);
  const json = await r.loadJSON()

  let listwidget = false
  let iterations = 0;
  do {
    if (iterations >= 1) {
      return makeErrorNotice();
    }

    listwidget = await attemptWidgetCreation(json)
    iterations++
  } while (listwidget === false)

  return listwidget;
}

async function attemptWidgetCreation(json) {
  try {
    let listwidget = new ListWidget();
    listwidget.url = "https://wanikani.com"

    if (json.leech_lesson_items.length === 0) {
      return makeNoLeechNotice();
    }

    const index = Math.floor(Math.random() * (json.leech_lesson_items.length - 1));
    const type = json.leech_lesson_items[index].type

    switch (type) {
      case "radical":
        listwidget.backgroundColor = new Color(radicalColor);
        await buildRadicalText(listwidget, json.leech_lesson_items[index]);
        break;
      case "kanji":
        listwidget.backgroundColor = new Color(kanjiColor);
        await buildKanjiText(listwidget, json.leech_lesson_items[index]);
        break;
      case "vocabulary":
        listwidget.backgroundColor = new Color(vocabColor);
        await buildVocabText(listwidget, json.leech_lesson_items[index]);
        break;
      default:
    }

    return listwidget
  } catch (e) {
    console.log(e);
    return false;
  }
}

function makeApiTokenNotice() {
  let listwidget = new ListWidget();
  listwidget.backgroundColor = new Color("#ffaa01");

  let notice = listwidget.addText("API Token is not set.");
  notice.centerAlignText();
  notice.font = Font.blackRoundedSystemFont(smallSize)
  notice.textColor = new Color("#ffffff");
  shadow(notice, smallSize, "#dd9200");

  return listwidget;
}

function makeNoLeechNotice() {
  const color = "#008343";
  const shadowColor = "#006333";
  let listwidget = new ListWidget();
  listwidget.backgroundColor = new Color(color);

  let notice = listwidget.addText("Good job genius! ❤︎ You have no leeches!");
  notice.centerAlignText();
  notice.font = Font.blackRoundedSystemFont(smallSize)
  notice.textColor = new Color("#ffffff");
  shadow(notice, smallSize, shadowColor);

  return listwidget;
}

function makeErrorNotice() {
  let listwidget = new ListWidget();
  listwidget.backgroundColor = new Color("#ee2201");

  let notice = listwidget.addText("Oops! Something went wrong...");
  notice.centerAlignText();
  notice.font = Font.blackRoundedSystemFont(smallSize)
  notice.textColor = new Color("#ffffff");
  shadow(notice, smallSize, "#922201");

  return listwidget;
}

async function buildRadicalText(listwidget, json) {
  const type = json.leech.key.split('/')[0]
  const key = encodeURIComponent(json.leech.key.split('/')[1])
  listwidget.url = `https://wanikani.com/${type}/${key}`


  let target = listwidget.addText(json.name);
  target.centerAlignText();
  target.font = Font.blackRoundedSystemFont(bigSize)
  target.textColor = new Color(radicalTextColor);


  listwidget.addSpacer(spacerSize);

  let meaning = listwidget.addText(json.correct_answers.join(', '));
  meaning.centerAlignText();
  meaning.font = Font.lightSystemFont(smallSize);
  meaning.textColor = new Color(radicalTextColor);
  shadow(meaning, smallSize, radicalShadowColor);
}

async function buildKanjiText(listwidget, json) {
  const type = json.leech.key.split('/')[0]
  const key = encodeURIComponent(json.leech.key.split('/')[1])
  listwidget.url = `https://wanikani.com/${type}/${key}`

  const { reading, meaning } = await scrapeKanji(key);

  let target = listwidget.addText(json.name);
  target.centerAlignText();
  target.font = Font.blackRoundedSystemFont(bigSize)
  target.textColor = new Color(kanjiTextColor);
  shadow(target, bigSize, kanjiShadowColor);

  listwidget.addSpacer(spacerSize);

  let rt = listwidget.addText(reading);
  rt.centerAlignText();
  rt.font = Font.lightSystemFont(smallSize);
  rt.textColor = new Color(kanjiTextColor);
  let mt = listwidget.addText(meaning);
  mt.centerAlignText();
  mt.font = Font.lightSystemFont(smallSize);
  mt.textColor = new Color(kanjiTextColor);
  shadow(mt, smallSize, kanjiShadowColor);
  shadow(rt, smallSize, kanjiShadowColor);
}

async function buildVocabText(listwidget, json) {
  const type = json.leech.key.split('/')[0]
  const key = encodeURIComponent(json.leech.key.split('/')[1])
  listwidget.url = `https://wanikani.com/${type}/${key}`

  const { reading, meaning } = await scrapeVocab(key);

  let fSize = bigSize;
  if (json.name.length !== 1) fSize = bigSize - ((json.name.length) * 4);
  if (fSize < smallSize) fSize = smallSize;
  let target = listwidget.addText(json.name);
  target.centerAlignText();
  target.font = Font.blackRoundedSystemFont(fSize)
  target.textColor = new Color(vocabTextColor);
  shadow(target, fSize, vocabShadowColor);

  listwidget.addSpacer(spacerSize);

  let rt = listwidget.addText(reading);
  rt.centerAlignText();
  rt.font = Font.lightSystemFont(smallSize);
  rt.textColor = new Color(vocabTextColor);
  let mt = listwidget.addText(meaning);
  mt.centerAlignText();
  mt.font = Font.lightSystemFont(smallSize);
  mt.textColor = new Color(vocabTextColor);
  shadow(mt, smallSize, vocabShadowColor);
  shadow(rt, smallSize, vocabShadowColor);
}

async function scrapeVocab(item) {
  const url = `https://www.wanikani.com/vocabulary/${item}`;
  let r = new Request(url);
  const data = await r.loadString()

  try {
    const meaningChunk = data.split('<h2>Primary</h2>')[1]
    const meaning = meaningChunk.split('</div>')[0].replace(/<[^>]*>?/gm, '').trimStart().trimEnd();
    const readingChunk = meaningChunk.split('<p class="pronunciation-variant" lang="ja">')[1];
    const reading = readingChunk.split('</p>')[0].replace(/<[^>]*>?/gm, '').trimStart().trimEnd();
    return {
      meaning,
      reading
    }
  } catch (e) {
    console.log(data);
    console.log(e);
    console.log("vocabulary/" + item);
    throw e;
  }
}

async function scrapeKanji(item) {
  const url = `https://www.wanikani.com/kanji/${item}`;
  let r = new Request(url);
  const data = await r.loadString()
  try {
    const meaningChunk = data.split('<h2>Primary</h2>')[1]
    const meaning = meaningChunk.split('</div>')[0].replace(/<[^>]*>?/gm, '').trimStart().trimEnd();
    const preReadingChunk = meaningChunk.split(`<div class="span4 ">`)[1];
    const readingChunk = preReadingChunk.split(`<p lang="ja">`)[1];
    const reading = readingChunk.split('</p>')[0].replace(/<[^>]*>?/gm, '').trimStart().trimEnd();
    return {
      meaning,
      reading
    }
  } catch (e) {
    console.log(data);
    console.log(e);
    console.log("kanji/" + item);
    throw e;
  }
}

function shadow(target, size, color) {
  p = size * (3 / 32);

  target.shadowColor = new Color(color);
  target.shadowOffset = new Point(p, p);
  target.shadowRadius = 0.1;
}
